Drafting SL ref docs (#8709)
fishtown-analytics/docs.getdbt.com: website/docs/reference/semantic-layer-reference.md

id: semantic-layer-reference
description: "Configuration reference for the Semantic Layer: semantic models, metrics, and dimensions."
keywords:
  - dbt semantic layer
  - semantic layer configuration
  - metrics configuration
  - dimensions configuration
---

## Related documentation

- [Build your metrics](/docs/build/build-metrics-intro): Conceptual overview of the Semantic Layer, metric types, and how to get started.
- [Semantic models](/docs/build/semantic-models): How to define semantic models (standalone or on a model), structure, and examples.
- [Creating metrics](/docs/build/metrics-overview): How to create and configure metrics; links to type-specific guides (simple, cumulative, ratio, derived, conversion).
- [Dimensions](/docs/build/dimensions): How to define time and categorical dimensions within semantic models.

This section is the **configuration reference** for the Semantic Layer. It lists every property and option for semantic models, metrics, and dimensions so you can see the full spec in one place.

## Property reference

Click the link on each option to read the full property reference.

| Page | Description |
|------|-------------|

> | [Semantic model properties](/reference/semantic-model-properties) | Properties for semantic models (entities, measures, dimensions, defaults). |

Reviewer: Measures are deprecated in the new spec
```suggestion
| [Semantic model properties](/reference/semantic-model-properties) | Properties for semantic models (entities, <VersionBlock lastVersion="1.11">measures</VersionBlock><VersionBlock firstVersion="1.12">simple metrics</VersionBlock>, dimensions, defaults). |
```

| [Metric properties](/reference/metric-properties) | Properties for metrics (simple, cumulative, ratio, derived, conversion), including metrics defined on a model. |
| [Dimension properties](/reference/dimension-properties) | Properties for dimensions (time and categorical). |

## Where to define Semantic Layer objects

You can define Semantic Layer objects in the following ways:

> - **Semantic models** — In standalone YAML or in <Constant name="fusion" /> as a `semantic_model` block on a [model](/reference/model-properties).

Reviewer: ```suggestion
- **Semantic models** — In standalone YAML, or as a `semantic_model` block under [`models:`](/reference/model-properties) in the latest YAML spec.
```

> - **Metrics** — In standalone YAML or in <Constant name="fusion" /> under a semantic model on a model (`semantic_model.metrics`).

Reviewer: ```suggestion
- **Metrics** — In standalone YAML or as a metrics list on the model (alongside `semantic_model`) in the latest YAML spec.
```

import LatestYamlSpecAvailability from '/snippets/_latest-yaml-spec-availability.md';

<LatestYamlSpecAvailability />

## Property reference

The property reference pages document each resource type in detail so you can look up allowed values, syntax, and behavior for every property in the full YAML spec. Use the version picker on each page to see *latest spec (model YAML)* or *legacy spec (standalone YAML)* content:

- [Semantic model properties](/reference/semantic-model-properties)
- [Metric properties](/reference/metric-properties)
- [Dimension properties](/reference/dimension-properties)

## Where to define <Constant name="semantic_layer" /> objects

| Object | Latest spec (model YAML) | Legacy spec (standalone YAML) |
|--------|--------------------------|------------------------------|
| **Semantic models** | Top-level `semantic_model:` block under [models](/reference/model-properties). | Top-level `semantic_models:` list. |

> | **Metrics** | For metrics that only depend on the same semantic model, list them alongside `semantic_model:` and `columns:` (not nested under semantic_model). <br/ > <br/ > For metrics that depend on metrics or dimensions from a different semantic model, define them under a top-level `metrics:` block (For example, `outside models:`). This can be in the same YAML file or a separate one. | Top-level `metrics:` key in standalone YAML. |

Reviewer 2: 👍

keywords:
  - dbt semantic layer
  - semantic layer configuration
  - metrics configuration
  - dimensions configuration
---

You can use this <Constant name="semantic_layer" /> configuration reference to learn about every property and option for semantic models, metrics, and dimensions that belong to the full YAML spec.

Because this reference provides information about both the *latest spec* (model-embedded) and the *legacy spec* (standalone YAML), you'll need to select the appropriate version from the version picker. [Read the build docs](/docs/build/semantic-models) to find out which applies to your environment. To convert from the legacy spec, see [Migrate to the latest YAML spec](/docs/build/latest-metrics-spec).

import LatestYamlSpecAvailability from '/snippets/_latest-yaml-spec-availability.md';

<LatestYamlSpecAvailability />

## Property reference

Click a link to read the full property reference for that resource. Use the version picker on each page to see *latest spec (model YAML)* or *legacy spec (standalone YAML)* content.

| Page | Description |
|------|-------------|

> | [Semantic model properties](/reference/semantic-model-properties) | **Semantic models** — Latest spec (model YAML): `semantic_model:` block, top-level `agg_time_dimension` and `primary_entity`, `columns` (entity/dimension blocks, `granularity`), `derived_semantics`, and optional `metrics`. Legacy spec (standalone YAML): `semantic_models:` list with model, defaults, entities, dimensions, measures, config. |

Reviewer: The values in the second column are a bit difficult to read. Maybe we can use `<br></br>` so they're not in one line? For example:
```suggestion
| [Semantic model properties](/reference/semantic-model-properties) | **Semantic models**: <br></br>- Latest spec (model YAML): `semantic_model:` block, top-level `agg_time_dimension` and `primary_entity`, `columns` (entity/dimension blocks, `granularity`), `derived_semantics`, and optional `metrics`.<br></br> - Legacy spec (standalone YAML): `semantic_models:` list with model, defaults, entities, dimensions, measures, config. |
```

| [Metric properties](/reference/metric-properties) | **Metrics** — Latest spec (model YAML): top-level `metrics:` list on a model; property table and properties by type (simple, cumulative, ratio, derived, conversion); cross-model rules. Legacy spec (standalone YAML): top-level `metrics:` key; property table with `type_params`; type-specific parameters. |
| [Dimension properties](/reference/dimension-properties) | **Dimensions** — Latest spec (model YAML): column-level `dimension:` block, `granularity:` on column, `derived_semantics.dimensions`, `validity_params`. Legacy spec (standalone semantic model): `dimensions:` list with name, type, `type_params`, description, expr, label, meta. |

## Where to define <Constant name="semantic_layer" /> objects

| Object | Latest spec (model YAML) | Legacy spec (standalone YAML) |
|--------|--------------------------|------------------------------|
| **Semantic models** | Top-level `semantic_model:` block on a [model](/reference/model-properties). | Top-level `semantic_models:` list. |

> | **Metrics** | Top-level `metrics:` list on a model alongside `semantic_model:` and `columns:` (not nested under `semantic_model`). Model-level `metrics:` can only reference that semantic model; for cross-model metrics, use standalone YAML. | Top-level `metrics:` key in standalone YAML. |

Reviewer: The latest spec description for Metrics is a bit confusing, should we use:

For latest spec:
Model-level `metrics:` list alongside `semantic_model:` and `columns:` (not nested under `semantic_model`). Use this for metrics that only depend on the same semantic model. For metrics that depend on metrics or dimensions from a different semantic model, define them under a top-level `metrics:` block (outside models:). This can be in the same YAML file or a separate one.

> | **Dimensions** | `dimension:` blocks on model columns (and optional `derived_semantics.dimensions`). Always within a semantic model. | `dimensions:` list on the semantic model. Always within a semantic model. |

Reviewer: Saying "always within a semantic model" under the latest spec might mean that it's under the `semantic_model` block? Maybe we can use "Defined within a model's semantic layer configuration." instead

Author: This is really helpful thanks @luna-bianca!
